fastlane_core/lib/fastlane_core/itunes_transporter.rb

```diff
@@ -180,9 +180,124 @@ def additional_upload_parameters
     end
   end
 
+  # Generates commands and executes the altool.
+  class AltoolTransporterExecutor < TransporterExecutor
+    ERROR_REGEX = /\sNSLocalizedFailureReason\s=\s"+(.+)"/
+
+    private_constant :ERROR_REGEX
+
+    def execute(command, hide_output)
+      if Helper.test?
+        yield(nil) if block_given?
+        command
+      end
+
+      @errors = []
+      @warnings = []
+      @all_lines = []
+
+      if hide_output
+        # Show a one time message instead
+        UI.success("Waiting for App Store Connect transporter to be finished.")
+        UI.success("Application Loader progress... this might take a few minutes...")
+      end
+
+      begin
+        exit_status = FastlaneCore::FastlanePty.spawn(command) do |command_stdout, command_stdin, pid|
+          begin
+            command_stdout.each do |line|
+              @all_lines << line
+              parse_line(line, hide_output) # this is where the parsing happens
+            end
+          end
+        end
+      rescue => ex
+        # FastlanePty adds exit_status on to StandardError so every error will have a status code
+        exit_status = ex.exit_status
+        @errors << ex.to_s
+      end
+
+      @errors << "The call to the altool completed with a non-zero exit status: #{exit_status}. This indicates a failure." unless exit_status.zero?
+
+      UI.important(@warnings.join("\n")) unless @warnings.empty?
+
+      if @errors.count > 0 && @all_lines.count > 0
+        # Print out the last 18 lines, this is key for non-verbose mode
+        @all_lines.last(18).each do |line|
```

Comment:
Are we sure this will always be the last 18 lines? What if there are other metadata in the JSON printed in `altool`'s log that makes in take more or less lines, depending on the type of error and the `original_server_error`, or if the `NSLocalizedDescription`, `NSLocalizedFailureReason`, and/or `title` ends up being some multiline text depending on the error you encounter?

Reply:
> Are we sure this will always be the last 18 lines?

Honestly, I'm not sure 😅 . I only encountered and know `original_server_error ` error and it `almost` 18 line to show JSON.

Reply:
What about detecting the first line with `*** Error:` in it, and print all the lines from that one to the end?

Reply:
Or alternatively, detecting each line starting with `/[0-9 .:-]* \*\*\* Error:/`, and for each one, accumulate all the following lines *not* starting with a timestamp like `/[0-9 .:-]*/` — to pick up all lines from that error… up to the next timestamped log line, if any 🤔

Reply:
https://github.com/fastlane/fastlane/pull/20631/commits/01a58737a9a805b3d8034a63f8c94d7725a992e0 
I wanna propose this improvement. This will shows the lines until `***Error` shown. If `***Error` cannot be found (it is super very rare case since `@errors.count > 0 && @all_lines.count > 0`), shows last 20 line

Reply:
Sample is like below

```
[08:25:51]: This might take a few minutes. Please don't interrupt the script.
[08:25:55]: [altool] 2022-09-11 08:25:55.615 *** Error: Error uploading '/xxx.pkg'.

[08:25:55]: [altool] 2022-09-11 08:25:55.615 *** Error: The provided entity includes an attribute with a value that has already been used The bundle version must be higher than the previously uploaded version: ‘3’. (ID: xxx) (-19232)

[08:25:55]: [altool]  {

[08:25:55]: [altool]     NSLocalizedDescription = "The provided entity includes an attribute with a value that has already been used";

[08:25:55]: [altool]     NSLocalizedFailureReason = "The bundle version must be higher than the previously uploaded version: \U20183\U2019. (ID: xxx)";

[08:25:55]: [altool]     "original_server_error" =     {

[08:25:55]: [altool]         code = "ENTITY_ERROR.ATTRIBUTE.INVALID.DUPLICATE";

[08:25:55]: [altool]         detail = "The bundle version must be higher than the previously uploaded version.";

[08:25:55]: [altool]         id = "xxx";

[08:25:55]: [altool]         meta =         {

[08:25:55]: [altool]             previousBundleVersion = 3;

[08:25:55]: [altool]         };

[08:25:55]: [altool]         source =         {

[08:25:55]: [altool]             pointer = "/data/attributes/cfBundleVersion";

[08:25:55]: [altool]         };

[08:25:55]: [altool]         status = 409;

[08:25:55]: [altool]         title = "The provided entity includes an attribute with a value that has already been used";

[08:25:55]: [altool]     };

[08:25:55]: [altool]     previousBundleVersion = 3;

[08:25:55]: [altool] }
```

Reply:
Other sample is below which has less line log

```
[08:23:17]: This might take a few minutes. Please don't interrupt the script.
[08:23:32]: [altool] 2022-09-11 08:23:32.551 *** Error: Error uploading '/xxx.pkg'.

[08:23:32]: [altool] 2022-09-11 08:23:32.552 *** Error: Asset validation failed The product archive is invalid. The Info.plist must contain a LSApplicationCategoryType key, whose value is the UTI for a valid category. For more details, see "Submitting your Mac apps to the App Store". (ID: xxx) (90242)

[08:23:32]: [altool]  {

[08:23:32]: [altool]     NSLocalizedDescription = "Asset validation failed";

[08:23:32]: [altool]     NSLocalizedFailureReason = "The product archive is invalid. The Info.plist must contain a LSApplicationCategoryType key, whose value is the UTI for a valid category. For more details, see \"Submitting your Mac apps to the App Store\". (ID: xxx)";

[08:23:32]: [altool]     "original_server_error" =     {

[08:23:32]: [altool]         code = "STATE_ERROR.VALIDATION_ERROR.90242";

[08:23:32]: [altool]         detail = "The product archive is invalid. The Info.plist must contain a LSApplicationCategoryType key, whose value is the UTI for a valid category. For more details, see \"Submitting your Mac apps to the App Store\".";

[08:23:32]: [altool]         id = "xxxx";

[08:23:32]: [altool]         status = 409;

[08:23:32]: [altool]         title = "Asset validation failed";

[08:23:32]: [altool]     };

[08:23:32]: [altool] }

```

Reply:
I like the idea 👍

Note that in terms of implementation you can do some optimisations, like using `@all_lines.rindex { … }` to find the index of the last line matching the regex (instead of using `each_index { … }.last`.
(Or you might be able to use `drop_while` or similar? Though that one might not get you the last occurrence)

Reply:
Ohhhh! I need to learn more about Ruby, Thank you. Let me fix

I feel that `rindex` is suitable for my case since needs to check from bottom

Reply:
fixed! https://github.com/fastlane/fastlane/pull/20631/commits/7490c5c0742d7e3eba540b3a7b61c9a4f70bffda Thank you so much!

Reply:
Another tip: you can [index arrays using ranges and negative indices](https://ruby-doc.org/core-3.1.2/Array.html#method-i-5B-5D) like `arr[-20..-1]` for example (equivalent to `last(20)`) or `arr[error_line..-1]` (to get lines from index `error_line` to the end of the array)

I think using that instead of `last(…your long condition + math…)` would make the code more readable.

So this would become something like `@all_lines[(error_line || -20)..-1].each do … end`

Reply:
ohhh.. cool!

Reply:
Also you should add those example messages you commented above … as unit tests 🙂

Reply:
good. let me add it later 🙇 (Sorry I'm busy from now so it will be later)

Reply:
No worries, and thanks already for handling all this 👍